### ISSUE TITLE
Duplicated code

### ISSUE THREAD

**Issue by samreid** [phetsims/states-of-matter#188]
While looking into https://github.com/phetsims/chipper/issues/526 I noticed that states-of-matter had around 280 lines of duplicated code.  It seemed higher than I would expect, so I thought I would make a note of it.  Here is the subset of the report from the chipper issue that points out the duplicated code in states-of-matter:

```
	- ../fluid-pressure-and-flow/js/flow/view/FlowView.js: 229-244
	 ../states-of-matter/js/atomic-interactions/view/AtomicInteractionsScreenView.js: 197-212

 229 │ 197 │       ( slowMotionRadioBox.localBounds.minX + speedControlMaxWidth ),   
 230 │ 198 │       slowMotionRadioBox.localBounds.maxY + touchAreaYDilation          
 231 │ 199 │     );                                                                  
 232 │ 200 │                                                                         
 233 │ 201 │     normalMotionRadioBox.touchArea = new Bounds2(                       
 234 │ 202 │       normalMotionRadioBox.localBounds.minX,                            
 235 │ 203 │       normalMotionRadioBox.localBounds.minY - touchAreaYDilation,       
 236 │ 204 │       ( normalMotionRadioBox.localBounds.minX + speedControlMaxWidth ), 
 237 │ 205 │       normalMotionRadioBox.localBounds.maxY + touchAreaYDilation        
 238 │ 206 │     );                                                                  
 239 │ 207 │                                                                         
 240 │ 208 │     var speedControl = new VBox( {                                      
 241 │ 209 │       align: 'left',                                                    
 242 │ 210 │       spacing: radioButtonSpacing,                                      
 243 │ 211 │       children: [ slowMotionRadioBox, normalMotionRadioBox ]            
 244 │ 212 │     } );                                                                

	- ../states-of-matter/js/atomic-interactions/view/InteractiveInteractionPotentialDiagram.js: 74-85
	 ../states-of-matter/js/atomic-interactions/view/InteractiveInteractionPotentialDiagram.js: 115-126

 74 │ 115 │         endDragY = self.epsilonLine.globalToParentPoint( event.pointer.point ).y;                           
 75 │ 116 │         var d = endDragY - startDragY;                                                                      
 76 │ 117 │         startDragY = endDragY;                                                                              
 77 │ 118 │         var scaleFactor = StatesOfMatterConstants.MAX_EPSILON / ( self.getGraphHeight() / 2 );              
 78 │ 119 │         dualAtomModel.interactionStrengthProperty.value = dualAtomModel.getEpsilon() + ( d * scaleFactor ); 
 79 │ 120 │       },                                                                                                    
 80 │ 121 │                                                                                                             
 81 │ 122 │       end: function() {                                                                                     
 82 │ 123 │         dualAtomModel.setMotionPaused( false );                                                             
 83 │ 124 │       }                                                                                                     
 84 │ 125 │     } ) );                                                                                                  
 85 │ 126 │     this.epsilonLineLayer.addChild( this.epsilonLine );                                                     

	- ../states-of-matter/js/common/model/engine/DiatomicPhaseStateChanger.js: 149-162
	 ../states-of-matter/js/common/model/engine/DiatomicPhaseStateChanger.js: 226-239

 149 │ 226 │       this.multipleParticleModel.setTemperature( StatesOfMatterConstants.LIQUID_TEMPERATURE );         
 150 │ 227 │                                                                                                        
 151 │ 228 │       // Get references to the various elements of the data set.                                       
 152 │ 229 │       var moleculeDataSet = this.multipleParticleModel.moleculeDataSet;                                
 153 │ 230 │       var moleculeCenterOfMassPositions = moleculeDataSet.getMoleculeCenterOfMassPositions();          
 154 │ 231 │       var moleculeVelocities = moleculeDataSet.getMoleculeVelocities();                                
 155 │ 232 │       var moleculeRotationAngles = moleculeDataSet.getMoleculeRotationAngles();                        
 156 │ 233 │       var moleculeRotationRates = moleculeDataSet.getMoleculeRotationRates();                          
 157 │ 234 │       var moleculesInsideContainer = this.multipleParticleModel.moleculeDataSet.insideContainer;       
 158 │ 235 │                                                                                                        
 159 │ 236 │       // Create and initialize other variables needed to do the job.                                   
 160 │ 237 │       var temperatureSqrt = Math.sqrt( this.multipleParticleModel.temperatureSetPointProperty.get() ); 
 161 │ 238 │       var numberOfMolecules = moleculeDataSet.getNumberOfMolecules();                                  
 162 │ 239 │                                                                                                        

	- ../states-of-matter/js/common/model/engine/DiatomicPhaseStateChanger.js: 263-273
	 ../states-of-matter/js/common/model/engine/MonatomicPhaseStateChanger.js: 223-233

 263 │ 223 │       for ( i = 0; i < numberOfMolecules; i++ ) {                                                            
 264 │ 224 │         for ( var j = 0; j < this.MAX_PLACEMENT_ATTEMPTS; j++ ) {                                            
 265 │ 225 │                                                                                                              
 266 │ 226 │           // Pick a random position.                                                                         
 267 │ 227 │           newPosX = this.MIN_INITIAL_PARTICLE_TO_WALL_DISTANCE + ( phet.joist.random.nextDouble() * rangeX); 
 268 │ 228 │           newPosY = this.MIN_INITIAL_PARTICLE_TO_WALL_DISTANCE + ( phet.joist.random.nextDouble() * rangeY); 
 269 │ 229 │           var positionAvailable = true;                                                                      
 270 │ 230 │                                                                                                              
 271 │ 231 │           // See if this position is available.                                                              
 272 │ 232 │           for ( var k = 0; k < i; k++ ) {                                                                    
 273 │ 233 │             if ( moleculeCenterOfMassPositions[ k ].distanceXY( newPosX, newPosY ) <                         

	- ../states-of-matter/js/common/model/engine/MonatomicPhaseStateChanger.js: 36-58
	 ../states-of-matter/js/common/model/engine/WaterPhaseStateChanger.js: 47-69

 36 │ 47 │   return inherit( AbstractPhaseStateChanger, MonatomicPhaseStateChanger, { 
 37 │ 48 │                                                                            
 38 │ 49 │     /**                                                                    
 39 │ 50 │      * @public                                                             
 40 │ 51 │      * @param {number} phaseID - state(solid/liquid/gas) of Molecule       
 41 │ 52 │      */                                                                    
 42 │ 53 │     setPhase: function( phaseID ) {                                        
 43 │ 54 │       switch( phaseID ) {                                                  
 44 │ 55 │         case PhaseStateEnum.SOLID:                                         
 45 │ 56 │           this.setPhaseSolid();                                            
 46 │ 57 │           break;                                                           
 47 │ 58 │         case PhaseStateEnum.LIQUID:                                        
 48 │ 59 │           this.setPhaseLiquid();                                           
 49 │ 60 │           break;                                                           
 50 │ 61 │         case PhaseStateEnum.GAS:                                           
 51 │ 62 │           this.setPhaseGas();                                              
 52 │ 63 │           break;                                                           
 53 │ 64 │         default:                                                           
 54 │ 65 │           throw new Error( 'invalid phaseID: ' + phaseID );                
 55 │ 66 │       }                                                                    
 56 │ 67 │                                                                            
 57 │ 68 │       var moleculeDataSet = this.multipleParticleModel.moleculeDataSet;    
 58 │ 69 │                                                                            

	- ../states-of-matter/js/common/model/engine/DiatomicPhaseStateChanger.js: 161-171
	 ../states-of-matter/js/common/model/engine/WaterPhaseStateChanger.js: 100-110

 161 │ 100 │       var numberOfMolecules = moleculeDataSet.getNumberOfMolecules();                                                            
 162 │ 101 │                                                                                                                                  
 163 │ 102 │       // Initialize the velocities and angles of the molecules.                                                                  
 164 │ 103 │       for ( var i = 0; i < numberOfMolecules; i++ ) {                                                                            
 165 │ 104 │                                                                                                                                  
 166 │ 105 │         // Assign each molecule an initial velocity.                                                                             
 167 │ 106 │         moleculeVelocities[ i ].setXY( temperatureSqrt * this.rand.nextGaussian(), temperatureSqrt * this.rand.nextGaussian() ); 
 168 │ 107 │                                                                                                                                  
 169 │ 108 │         // Assign each molecule an initial rotation rate.                                                                        
 170 │ 109 │         moleculeRotationRates[ i ] = phet.joist.random.nextDouble() * temperatureSqrt * Math.PI * 2;                             
 171 │ 110 │                                                                                                                                  

	- ../states-of-matter/js/common/model/engine/DiatomicPhaseStateChanger.js: 160-171
	 ../states-of-matter/js/common/model/engine/WaterPhaseStateChanger.js: 160-171

 160 │ 160 │       var temperatureSqrt = Math.sqrt( this.multipleParticleModel.temperatureSetPointProperty.get() );                           
 161 │ 161 │       var numberOfMolecules = moleculeDataSet.getNumberOfMolecules();                                                            
 162 │ 162 │                                                                                                                                  
 163 │ 163 │       // Initialize the velocities and angles of the molecules.                                                                  
 164 │ 164 │       for ( var i = 0; i < numberOfMolecules; i++ ) {                                                                            
 165 │ 165 │                                                                                                                                  
 166 │ 166 │         // Assign each molecule an initial velocity.                                                                             
 167 │ 167 │         moleculeVelocities[ i ].setXY( temperatureSqrt * this.rand.nextGaussian(), temperatureSqrt * this.rand.nextGaussian() ); 
 168 │ 168 │                                                                                                                                  
 169 │ 169 │         // Assign each molecule an initial rotation rate.                                                                        
 170 │ 170 │         moleculeRotationRates[ i ] = phet.joist.random.nextDouble() * temperatureSqrt * Math.PI * 2;                             
 171 │ 171 │                                                                                                                                  

	- ../states-of-matter/js/common/model/engine/DiatomicPhaseStateChanger.js: 178-202
	 ../states-of-matter/js/common/model/engine/WaterPhaseStateChanger.js: 179-203

 178 │ 179 │       var centerPointY = this.multipleParticleModel.normalizedContainerHeight / 4;                                        
 179 │ 180 │       var currentLayer = 0;                                                                                               
 180 │ 181 │       var particlesOnCurrentLayer = 0;                                                                                    
 181 │ 182 │       var particlesThatWillFitOnCurrentLayer = 1;                                                                         
 182 │ 183 │       for ( i = 0; i < numberOfMolecules; i++ ) {                                                                         
 183 │ 184 │         for ( var j = 0; j < this.MAX_PLACEMENT_ATTEMPTS; j++ ) {                                                         
 184 │ 185 │           var distanceFromCenter = currentLayer * MIN_INITIAL_DIAMETER_DISTANCE * LIQUID_SPACING_FACTOR;                  
 185 │ 186 │           var angle = (particlesOnCurrentLayer / particlesThatWillFitOnCurrentLayer * 2 * Math.PI) +                      
 186 │ 187 │                       (particlesThatWillFitOnCurrentLayer / (4 * Math.PI));                                               
 187 │ 188 │           var xPos = centerPointX + (distanceFromCenter * Math.cos( angle ));                                             
 188 │ 189 │           var yPos = centerPointY + (distanceFromCenter * Math.sin( angle ));                                             
 189 │ 190 │                                                                                                                           
 190 │ 191 │           // Consider this spot used even if we don't actually put the particle there.                                    
 191 │ 192 │           particlesOnCurrentLayer++;                                                                                      
 192 │ 193 │           if ( particlesOnCurrentLayer >= particlesThatWillFitOnCurrentLayer ) {                                          
 193 │ 194 │                                                                                                                           
 194 │ 195 │             // This layer is full - move to the next one.                                                                 
 195 │ 196 │             currentLayer++;                                                                                               
 196 │ 197 │             particlesThatWillFitOnCurrentLayer = Math.floor( currentLayer * 2 * Math.PI /                                 
 197 │ 198 │                                                              ( MIN_INITIAL_DIAMETER_DISTANCE * LIQUID_SPACING_FACTOR ) ); 
 198 │ 199 │             particlesOnCurrentLayer = 0;                                                                                  
 199 │ 200 │           }                                                                                                               
 200 │ 201 │           // Check if the position is too close to the wall.  Note                                                        
 201 │ 202 │           // that we don't check inter-particle distances here - we rely                                                  
 202 │ 203 │           // on the placement algorithm to make sure that this is not a                                                   

	- ../states-of-matter/js/common/model/engine/DiatomicPhaseStateChanger.js: 237-252
	 ../states-of-matter/js/common/model/engine/WaterPhaseStateChanger.js: 239-254

 237 │ 239 │       var temperatureSqrt = Math.sqrt( this.multipleParticleModel.temperatureSetPointProperty.get() );                           
 238 │ 240 │       var numberOfMolecules = moleculeDataSet.getNumberOfMolecules();                                                            
 239 │ 241 │       for ( var i = 0; i < numberOfMolecules; i++ ) {                                                                            
 240 │ 242 │                                                                                                                                  
 241 │ 243 │         // Temporarily position the molecules at (0,0).                                                                          
 242 │ 244 │         moleculeCenterOfMassPositions[ i ].setXY( 0, 0 );                                                                        
 243 │ 245 │                                                                                                                                  
 244 │ 246 │         // Assign each molecule an initial velocity.                                                                             
 245 │ 247 │         moleculeVelocities[ i ].setXY( temperatureSqrt * this.rand.nextGaussian(), temperatureSqrt * this.rand.nextGaussian() ); 
 246 │ 248 │                                                                                                                                  
 247 │ 249 │         // Assign each molecule an initial rotational position.                                                                  
 248 │ 250 │         moleculeRotationAngles[ i ] = phet.joist.random.nextDouble() * Math.PI * 2;                                              
 249 │ 251 │                                                                                                                                  
 250 │ 252 │         // Assign each molecule an initial rotation rate.                                                                        
 251 │ 253 │         moleculeRotationRates[ i ] = phet.joist.random.nextDouble() * temperatureSqrt * Math.PI * 2;                             
 252 │ 254 │                                                                                                                                  

	- ../states-of-matter/js/common/model/engine/DiatomicPhaseStateChanger.js: 262-286
	 ../states-of-matter/js/common/model/engine/WaterPhaseStateChanger.js: 264-288

 262 │ 264 │       var rangeY = this.multipleParticleModel.normalizedContainerWidth - ( 2 * this.MIN_INITIAL_PARTICLE_TO_WALL_DISTANCE); 
 263 │ 265 │       for ( i = 0; i < numberOfMolecules; i++ ) {                                                                           
 264 │ 266 │         for ( var j = 0; j < this.MAX_PLACEMENT_ATTEMPTS; j++ ) {                                                           
 265 │ 267 │                                                                                                                             
 266 │ 268 │           // Pick a random position.                                                                                        
 267 │ 269 │           newPosX = this.MIN_INITIAL_PARTICLE_TO_WALL_DISTANCE + ( phet.joist.random.nextDouble() * rangeX);                
 268 │ 270 │           newPosY = this.MIN_INITIAL_PARTICLE_TO_WALL_DISTANCE + ( phet.joist.random.nextDouble() * rangeY);                
 269 │ 271 │           var positionAvailable = true;                                                                                     
 270 │ 272 │                                                                                                                             
 271 │ 273 │           // See if this position is available.                                                                             
 272 │ 274 │           for ( var k = 0; k < i; k++ ) {                                                                                   
 273 │ 275 │             if ( moleculeCenterOfMassPositions[ k ].distanceXY( newPosX, newPosY ) <                                        
 274 │ 276 │                  MIN_INITIAL_DIAMETER_DISTANCE * GAS_SPACING_FACTOR ) {                                                     
 275 │ 277 │               positionAvailable = false;                                                                                    
 276 │ 278 │               break;                                                                                                        
 277 │ 279 │             }                                                                                                               
 278 │ 280 │           }                                                                                                                 
 279 │ 281 │           if ( positionAvailable ) {                                                                                        
 280 │ 282 │                                                                                                                             
 281 │ 283 │             // We found an open position.                                                                                   
 282 │ 284 │             moleculeCenterOfMassPositions[ i ].setXY( newPosX, newPosY );                                                   
 283 │ 285 │             break;                                                                                                          
 284 │ 286 │           }                                                                                                                 
 285 │ 287 │           else if ( j === this.MAX_PLACEMENT_ATTEMPTS - 1 ) {                                                               
 286 │ 288 │                                                                                                                             

	- ../states-of-matter/js/common/view/ParticleImageCanvasNode.js: 143-158
	 ../states-of-matter/js/common/view/ParticleImageCanvasNode.js: 162-177

 143 │ 162 │         if ( particle.renderBelowOxygen ){                                                                  
 144 │ 163 │           particleViewRadius = this.particleRadii[ particle.getType() ];                                    
 145 │ 164 │           context.drawImage(                                                                                
 146 │ 165 │             this.particleImageCanvas,                                                                       
 147 │ 166 │             this.mapAtomTypeToImageXPosition[ particle.getType() ],                                         
 148 │ 167 │             this.useStrokedParticles ? PARTICLE_IMAGE_CANVAS_LENGTH : 0,                                    
 149 │ 168 │             PARTICLE_IMAGE_CANVAS_LENGTH,                                                                   
 150 │ 169 │             PARTICLE_IMAGE_CANVAS_LENGTH,                                                                   
 151 │ 170 │             this.modelViewTransform.modelToViewX( particle.positionProperty.value.x ) - particleViewRadius, 
 152 │ 171 │             this.modelViewTransform.modelToViewY( particle.positionProperty.value.y ) - particleViewRadius, 
 153 │ 172 │             particleViewRadius * 2,                                                                         
 154 │ 173 │             particleViewRadius * 2                                                                          
 155 │ 174 │           );                                                                                                
 156 │ 175 │         }                                                                                                   
 157 │ 176 │       }                                                                                                     
 158 │ 177 │                                                                                                             

	- ../states-of-matter/js/atomic-interactions/view/InteractiveInteractionPotentialDiagram.js: 111-119
	 ../states-of-matter/js/phase-changes/view/EpsilonControlInteractionPotentialDiagram.js: 94-102

 111 │ 94  │         startDragY = self.epsilonResizeHandle.globalToParentPoint( event.pointer.point ).y;                 
 112 │ 95  │       },                                                                                                    
 113 │ 96  │                                                                                                             
 114 │ 97  │       drag: function( event ) {                                                                             
 115 │ 98  │         endDragY = self.epsilonResizeHandle.globalToParentPoint( event.pointer.point ).y;                   
 116 │ 99  │         var d = endDragY - startDragY;                                                                      
 117 │ 100 │         startDragY = endDragY;                                                                              
 118 │ 101 │         var scaleFactor = StatesOfMatterConstants.MAX_EPSILON / ( self.getGraphHeight() / 2 );              
 119 │ 102 │         dualAtomModel.interactionStrengthProperty.value = dualAtomModel.getEpsilon() + ( d * scaleFactor ); 

	- ../states-of-matter/js/atomic-interactions/view/InteractiveInteractionPotentialDiagram.js: 70-78
	 ../states-of-matter/js/phase-changes/view/EpsilonControlInteractionPotentialDiagram.js: 116-124

 70 │ 116 │         startDragY = self.epsilonLine.globalToParentPoint( event.pointer.point ).y;                         
 71 │ 117 │       },                                                                                                    
 72 │ 118 │                                                                                                             
 73 │ 119 │       drag: function( event ) {                                                                             
 74 │ 120 │         endDragY = self.epsilonLine.globalToParentPoint( event.pointer.point ).y;                           
 75 │ 121 │         var d = endDragY - startDragY;                                                                      
 76 │ 122 │         startDragY = endDragY;                                                                              
 77 │ 123 │         var scaleFactor = StatesOfMatterConstants.MAX_EPSILON / ( self.getGraphHeight() / 2 );              
 78 │ 124 │         dualAtomModel.interactionStrengthProperty.value = dualAtomModel.getEpsilon() + ( d * scaleFactor ); 

	- ../states-of-matter/js/phase-changes/view/PhaseChangesMoleculesControlPanel.js: 48-74
	 ../states-of-matter/js/states/view/StatesMoleculesControlPanel.js: 34-60

 48 │ 34 │   var ADJUSTABLE_ATTRACTION_ICON = new Circle( 6, { fill: StatesOfMatterConstants.ADJUSTABLE_ATTRACTION_COLOR } ); 
 49 │ 35 │                                                                                                                    
 50 │ 36 │   // icon for the neon button                                                                                      
 51 │ 37 │   var NEON_ICON = new Circle( 5, { fill: StatesOfMatterConstants.NEON_COLOR } );                                   
 52 │ 38 │                                                                                                                    
 53 │ 39 │   // icon for the argon button                                                                                     
 54 │ 40 │   var ARGON_ICON = new Circle( 6, { fill: StatesOfMatterConstants.ARGON_COLOR } );                                 
 55 │ 41 │                                                                                                                    
 56 │ 42 │   // icon for the water button                                                                                     
 57 │ 43 │   var dot1 = new Circle( 5, { fill: StatesOfMatterConstants.OXYGEN_COLOR } );                                      
 58 │ 44 │   var dot2 = new Circle( 3, {                                                                                      
 59 │ 45 │     fill: StatesOfMatterConstants.HYDROGEN_COLOR,                                                                  
 60 │ 46 │     stroke: StatesOfMatterConstants.HYDROGEN_COLOR, bottom: dot1.top + 5, right: dot1.left + 5                     
 61 │ 47 │   } );                                                                                                             
 62 │ 48 │   var dot3 = new Circle( 3, {                                                                                      
 63 │ 49 │     fill: StatesOfMatterConstants.HYDROGEN_COLOR,                                                                  
 64 │ 50 │     stroke: StatesOfMatterConstants.HYDROGEN_COLOR, bottom: dot1.top + 5, left: dot1.right - 5                     
 65 │ 51 │   } );                                                                                                             
 66 │ 52 │   var WATER_ICON = new Node( { children: [ dot3, dot1, dot2 ] } );                                                 
 67 │ 53 │                                                                                                                    
 68 │ 54 │   // icon for the oxygen button                                                                                    
 69 │ 55 │   var oxygen1 = new Circle( 5, { fill: StatesOfMatterConstants.OXYGEN_COLOR } );                                   
 70 │ 56 │   var oxygen2 = new Circle( 5, { fill: StatesOfMatterConstants.OXYGEN_COLOR, left: oxygen1.right - 4 } );          
 71 │ 57 │   var OXYGEN_ICON = new Node( { children: [ oxygen1, oxygen2 ] } );                                                
 72 │ 58 │                                                                                                                    
 73 │ 59 │   /**                                                                                                              
 74 │ 60 │    * @param { MultipleParticleModel } multipleParticleModel - model of the simulation                              

	- ../states-of-matter/js/phase-changes/view/PhaseChangesScreenView.js: 141-160
	 ../states-of-matter/js/states/view/StatesScreenView.js: 119-138

 141 │ 119 │       listener: function() { multipleParticleModel.stepInternal( 1 / 60 ); },             
 142 │ 120 │       radius: 12,                                                                         
 143 │ 121 │       stroke: 'black',                                                                    
 144 │ 122 │       fill: '#005566',                                                                    
 145 │ 123 │       right: heaterCoolerNode.left - STEP_BUTTON_X_OFFSET,                                
 146 │ 124 │       centerY: heaterCoolerNode.centerY,                                                  
 147 │ 125 │       touchAreaDilation: 4                                                                
 148 │ 126 │     } );                                                                                  
 149 │ 127 │     this.addChild( stepButton );                                                          
 150 │ 128 │                                                                                           
 151 │ 129 │     var playPauseButton = new PlayPauseButton( multipleParticleModel.isPlayingProperty, { 
 152 │ 130 │       radius: 18, stroke: 'black',                                                        
 153 │ 131 │       fill: '#005566',                                                                    
 154 │ 132 │       y: stepButton.centerY,                                                              
 155 │ 133 │       right: stepButton.left - 10,                                                        
 156 │ 134 │       touchAreaDilation: 4                                                                
 157 │ 135 │     } );                                                                                  
 158 │ 136 │     this.addChild( playPauseButton );                                                     
 159 │ 137 │                                                                                           
 160 │ 138 │     // add bicycle pump node                                                              

	- ../states-of-matter/js/phase-changes/view/PhaseChangesScreenView.js: 313-353
	 ../states-of-matter/js/states/view/StatesScreenView.js: 154-194

 313 │ 154 │       self.updatePhaseDiagram();                                                                            
 314 │ 155 │     } );                                                                                                    
 315 │ 156 │                                                                                                             
 316 │ 157 │     // if the appropriate query param is set, show some information used in debugging time step adjustments 
 317 │ 158 │     // TODO: Consider removing this once performance issues are worked out.                                 
 318 │ 159 │     if ( StatesOfMatterQueryParameters.debugTimeStep ) {                                                    
 319 │ 160 │       var keepingUpReadout = new Text( '', {                                                                
 320 │ 161 │         font: new PhetFont( 20 ),                                                                           
 321 │ 162 │         fill: 'red',                                                                                        
 322 │ 163 │         top: 40,                                                                                            
 323 │ 164 │         left: 20                                                                                            
 324 │ 165 │       } );                                                                                                  
 325 │ 166 │       this.addChild( keepingUpReadout );                                                                    
 326 │ 167 │       multipleParticleModel.keepingUpProperty.link( function( keepingUp ) {                                 
 327 │ 168 │         keepingUpReadout.text = keepingUp.toString();                                                       
 328 │ 169 │       } );                                                                                                  
 329 │ 170 │       var averageDtReadout = new Text( '', {                                                                
 330 │ 171 │         font: new PhetFont( 20 ),                                                                           
 331 │ 172 │         fill: 'red',                                                                                        
 332 │ 173 │         top: keepingUpReadout.bottom + 5,                                                                   
 333 │ 174 │         left: keepingUpReadout.left                                                                         
 334 │ 175 │       } );                                                                                                  
 335 │ 176 │       this.addChild( averageDtReadout );                                                                    
 336 │ 177 │       multipleParticleModel.averageDtProperty.link( function( averageDt ) {                                 
 337 │ 178 │         averageDtReadout.text = averageDt.toFixed( 3 );                                                     
 338 │ 179 │       } );                                                                                                  
 339 │ 180 │       var maxAdvanceTimePerStep = new Text( '', {                                                           
 340 │ 181 │         font: new PhetFont( 20 ),                                                                           
 341 │ 182 │         fill: 'red',                                                                                        
 342 │ 183 │         top: averageDtReadout.bottom + 5,                                                                   
 343 │ 184 │         left: averageDtReadout.left                                                                         
 344 │ 185 │       } );                                                                                                  
 345 │ 186 │       this.addChild( maxAdvanceTimePerStep );                                                               
 346 │ 187 │       multipleParticleModel.maxParticleMoveTimePerStepProperty.link( function( maxAdvance ) {               
 347 │ 188 │         maxAdvanceTimePerStep.text = maxAdvance.toFixed( 3 );                                               
 348 │ 189 │       } );                                                                                                  
 349 │ 190 │     }                                                                                                       
 350 │ 191 │   }                                                                                                         
 351 │ 192 │                                                                                                             
 352 │ 193 │   statesOfMatter.register( 'PhaseChangesScreenView', PhaseChangesScreenView );                              
 353 │ 194 │                                                                                                             
```

**Comment by jbphet** [comment]
The first item in @samreid's list above looks like 15 lines of duplicated code between states-of-matter and fluid-pressure-and-flow.  I'm not going to worry about this part as I work to tackle the other portions identified.

**Comment by jbphet** [comment]
40 of the lines were a result of some debug code which was just removed while addressing TODO items, see #189.  These were the lines identified as follows:

>
>	 ../states-of-matter/js/phase-changes/view/PhaseChangesScreenView.js: 313-353
>	 ../states-of-matter/js/states/view/StatesScreenView.js: 154-194

**Comment by jbphet** [comment]
I've gone through and eliminated most of the duplication.  I just re-ran the report, and it now looks like this (I omitted the overlap with FPAF, since I'm not going to look into any cross-repo duplication at this time):

```
	- ..\states-of-matter\js\common\model\engine\AbstractPhaseStateChanger.js: 201-211
	 ..\states-of-matter\js\common\model\engine\WaterPhaseStateChanger.js: 85-95

 201 │ 85 │       var numberOfMolecules = moleculeDataSet.getNumberOfMolecules();
 202 │ 86 │ 
 203 │ 87 │       // Initialize the velocities and angles of the molecules.
 204 │ 88 │       for ( var i = 0; i < numberOfMolecules; i++ ) {
 205 │ 89 │ 
 206 │ 90 │         // Assign each molecule an initial velocity.
 207 │ 91 │         moleculeVelocities[ i ].setXY( temperatureSqrt * this.rand.nextGaussian(), temperatureSqrt * this.rand.nextGaussian() );
 208 │ 92 │ 
 209 │ 93 │         // Assign each molecule an initial rotation rate.
 210 │ 94 │         moleculeRotationRates[ i ] = phet.joist.random.nextDouble() * temperatureSqrt * Math.PI * 2;
 211 │ 95 │ 

	- ..\states-of-matter\js\phase-changes\view\PhaseChangesScreenView.js: 139-158
	 ..\states-of-matter\js\states\view\StatesScreenView.js: 117-136

 139 │ 117 │       listener: function() { multipleParticleModel.stepInternal( 1 / 60 ); },
 140 │ 118 │       radius: 12,
 141 │ 119 │       stroke: 'black',
 142 │ 120 │       fill: '#005566',
 143 │ 121 │       right: heaterCoolerNode.left - STEP_BUTTON_X_OFFSET,
 144 │ 122 │       centerY: heaterCoolerNode.centerY,
 145 │ 123 │       touchAreaDilation: 4
 146 │ 124 │     } );
 147 │ 125 │     this.addChild( stepButton );
 148 │ 126 │ 
 149 │ 127 │     var playPauseButton = new PlayPauseButton( multipleParticleModel.isPlayingProperty, {
 150 │ 128 │       radius: 18, stroke: 'black',
 151 │ 129 │       fill: '#005566',
 152 │ 130 │       y: stepButton.centerY,
 153 │ 131 │       right: stepButton.left - 10,
 154 │ 132 │       touchAreaDilation: 4
 155 │ 133 │     } );
 156 │ 134 │     this.addChild( playPauseButton );
 157 │ 135 │ 
 158 │ 136 │     // add bicycle pump node
```

This amounts to 31 lines of duplicated code, and while I could probably generalize the phase state changer to eliminate the remain lines in the first group, and could generalize the inheritance hierarchy to eliminate those in the second group, it doesn't feel worth it to me.  In other words, I'm okay with where this is at now.  Back to @samreid to provide a second opinion as to whether this looks good and can be closed or whether he feels that more work is warranted.  He has looked at a number of repos, so he has a since of how much duplication generally arises.

**Comment by samreid** [comment]
In my opinion, this seems like a good opportunity to factor out code.  For the model, factoring out the code will help ensure that the AbstractPhaseStateChanger and WaterPhaseStateChanger don't accidentally diverge in the future (or if there are bugfixes they would only have to be done in one place.  In the view, factoring out the code will help ensure that the buttons maintain similar looks and styling if changed in the future.

For the model part, how about factoring out a function in AbstractPhaseStateChanger called initializeMoleculeVelocitiesAndAngles().

For the StepForwardButton and PlayPauseButton, I'd recommend to factor out StatesOfMatterStepForwardButton and StatesOfMatterPlayPauseButton which can be called from PhaseChangesScreenView and StatesScreenView.

**Comment by jbphet** [comment]
@samreid - I appreciate your high standards!  I've worked to consolidate the remaining duplicated code, and in the process it revealed an inconsistent behavior between the screens (the play pause button was inflating on one screen and not the others).  I re-ran the analyzer, and the only reported duplication was the aforementioned FPAF code (other than the expected config files).  I think our work here is done.  Closing.
